cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java

```diff
@@ -160,7 +195,8 @@ public ExecutionSparkContext(SparkSpecification specification, Id.Program progra
 
   @Override
   public synchronized <T extends Dataset> T getDataset(String name, Map<String, String> arguments) {
-    Map<String, String> datasetArgs = RuntimeArguments.extractScope(Scope.DATASET, name, getRuntimeArguments());
+    Map<String, String> datasetArgs = new HashMap<>();
+    datasetArgs.putAll(RuntimeArguments.extractScope(Scope.DATASET, name, getRuntimeArguments()));
     datasetArgs.putAll(arguments);
```

Comment:
Seems like the contract of the `RuntimeArguments.extractScope` is unclear. If the `getRuntimeArguments()` map passed to the method is an empty Map, then it returns the same map (which is immutable, since runtime args are immutable). But if there is something in the map, then it creates a HashMap and returns it.